File: ion/src/com/koushikdutta/ion/Loader.java
package com.koushikdutta.ion;

import android.content.Context;

import com.koushikdutta.async.DataEmitter;
import com.koushikdutta.async.future.Future;
import com.koushikdutta.async.future.FutureCallback;
import com.koushikdutta.async.http.AsyncHttpRequest;
import com.koushikdutta.ion.bitmap.BitmapInfo;
import com.koushikdutta.ion.future.ResponseFuture;

import java.io.InputStream;
import java.lang.reflect.Type;

/**
 * Created by koush on 5/22/13.
 */
public interface Loader {
    public static class LoaderEmitter {
        public LoaderEmitter(DataEmitter emitter, long length, ResponseServedFrom servedFrom,
                             HeadersResponse headers,
                             AsyncHttpRequest request) {
            this.length = length;
            this.emitter = emitter;
            this.servedFrom = servedFrom;
            this.headers = headers;
            this.request = request;
        }
        DataEmitter emitter;
        long length;
        public DataEmitter getDataEmitter() {
            return emitter;
        }
        public long length() {
            return length;
        }
        ResponseServedFrom servedFrom;
        public ResponseServedFrom getServedFrom() {
            return servedFrom;
        }
        HeadersResponse headers;
        public HeadersResponse getHeaders() {
            return headers;
        }
        AsyncHttpRequest request;
        public AsyncHttpRequest getRequest() {
            return request;
        }
    }

    /**
     * returns a Future if this loader can handle a request
     * otherwise it returns null, and Ion continues to the next loader.
     * @param ion
     * @param request
     * @param callback
     * @return
     */
    public Future<DataEmitter> load(Ion ion, AsyncHttpRequest request, FutureCallback<LoaderEmitter> callback);

    /**
     * returns a future if the laoder can handle the request as a bitmap
     * otherwise it returns null
     * @param ion
     * @param key
     * @param uri
     * @param resizeWidth
     * @param resizeHeight
     * @return
     */
    public Future<BitmapInfo> loadBitmap(Context context, Ion ion, String key, String uri, int resizeWidth, int resizeHeight, boolean animateGif);

    /**
     * Resolve a request into another request.
     * @param ion
     * @param request
     * @return
     */
    public Future<AsyncHttpRequest> resolve(Context context, Ion ion, AsyncHttpRequest request);

    public <T> ResponseFuture<T> load(Ion ion, AsyncHttpRequest request, Type type);
}

File: ion/src/com/koushikdutta/ion/Loader.kt
package com.koushikdutta.ion

import android.content.Context
import com.koushikdutta.ion.bitmap.BitmapInfo
import com.koushikdutta.ion.builder.ResponsePromise
import com.koushikdutta.scratch.AsyncInput
import com.koushikdutta.scratch.Promise
import com.koushikdutta.scratch.http.AsyncHttpRequest
import java.lang.reflect.Type

/**
 * Created by koush on 5/22/13.
 */
interface Loader {
    class LoaderResult(val input: AsyncInput, val length: Long?, val servedFrom: ResponseServedFrom,
                       val headers: HeadersResponse?,
                       val resolvedRequest: AsyncHttpRequest)

    /**
     * returns a Future if this loader can handle a request
     * otherwise it returns null, and Ion continues to the next loader.
     * @param ion
     * @param request
     * @param callback
     * @return
     */
    fun load(ion: Ion, request: AsyncHttpRequest): Promise<LoaderResult>?

    /**
     * returns a future if the laoder can handle the request as a bitmap
     * otherwise it returns null
     * @param ion
     * @param key
     * @param request
     * @param resizeWidth
     * @param resizeHeight
     * @return
     */
    fun loadBitmap(context: Context, ion: Ion, key: String, request: AsyncHttpRequest, resizeWidth: Int, resizeHeight: Int, animateGif: Boolean): Promise<BitmapInfo>?

    /**
     * Resolve a request into another request.
     * @param ion
     * @param request
     * @return
     */
    fun resolve(context: Context, ion: Ion, request: AsyncHttpRequest): Promise<AsyncHttpRequest>?

    /**
     * Resolve a request to a Response<T extends type>
     * @param ion
     * @param request
     * @param type
     * @return
     */
    fun <T> load(ion: Ion, request: AsyncHttpRequest, type: Type): ResponsePromise<T>?
}